File: app/src/main/java/com/dmi/perfectreader/book/pagebook/LocationUtils.java
package com.dmi.perfectreader.book.pagebook;

import com.dmi.util.lang.MathExt;

import static java8.util.J8Arrays.stream;

class LocationUtils {
    public static SegmentLocation percentToSegmentLocation(int[] segmentSizes, double percent) {
        int totalSize = stream(segmentSizes).sum();
        int position = (int) (percent * totalSize);
        int segmentStart, segmentEnd = 0;
        for (int i = 0; i < segmentSizes.length; i++) {
            segmentStart = segmentEnd;
            int segmentSize = segmentSizes[i];
            segmentEnd += segmentSize;
            if (position >= segmentStart && position < segmentEnd) {
                int positionInSegment = position - segmentStart;
                return new SegmentLocation(i, (double) positionInSegment / segmentSize);
            }
        }
        return new SegmentLocation(segmentSizes.length - 1, 1.0);
    }

    public static double segmentLocationToPercent(int[] segmentSizes, SegmentLocation segmentLocation) {
        int totalSize = stream(segmentSizes).sum();
        int position = 0;
        for (int i = 0; i < segmentLocation.index; i++) {
            position += segmentSizes[i];
        }
        position += segmentSizes[segmentLocation.index] * segmentLocation.percent;
        return MathExt.clamp(0.0, 1.0, position / totalSize);
    }

    public static int percentToPage(int pageCount, double percent) {
        return (int) Math.min(Math.round(percent * pageCount), pageCount - 1);
    }

    public static double pageToPercent(int pageCount, int page) {
        return (double) page / pageCount;
    }

    public static class SegmentLocation {
        public final int index;
        public final double percent;

        public SegmentLocation(int index, double percent) {
            this.index = index;
            this.percent = percent;
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/book/pagebook/LocationUtils.kt
package com.dmi.perfectreader.book.pagebook

import com.dmi.util.lang.MathExt

object LocationUtils {
    fun percentToSegmentLocation(segmentSizes: IntArray, percent: Double): SegmentLocation {
        val totalSize = segmentSizes.sum()
        val position = (percent * totalSize).toInt()
        var segmentStart: Int
        var segmentEnd = 0
        for (i in segmentSizes.indices) {
            segmentStart = segmentEnd
            val segmentSize = segmentSizes[i]
            segmentEnd += segmentSize
            if (position >= segmentStart && position < segmentEnd) {
                val positionInSegment = position - segmentStart
                return SegmentLocation(i, positionInSegment.toDouble() / segmentSize)
            }
        }
        return SegmentLocation(segmentSizes.size - 1, 1.0)
    }

    fun segmentLocationToPercent(segmentSizes: IntArray, segmentLocation: SegmentLocation): Double {
        val totalSize = segmentSizes.sum()
        var position = 0
        for (i in 0..segmentLocation.index - 1) {
            position += segmentSizes[i]
        }
        position += (segmentSizes[segmentLocation.index] * segmentLocation.percent).toInt()
        return MathExt.clamp(0.0, 1.0, (position / totalSize).toDouble())
    }

    fun percentToPage(pageCount: Int, percent: Double): Int {
        return Math.min(Math.round(percent * pageCount), (pageCount - 1).toLong()).toInt()
    }

    fun pageToPercent(pageCount: Int, page: Int): Double {
        return page.toDouble() / pageCount
    }

    class SegmentLocation(val index: Int, val percent: Double)
}
